src/cls/queue/cls_queue.cc

```diff
@@ -881,13 +886,42 @@ static int cls_gc_queue_list(cls_method_context_t hctx, bufferlist *in, bufferli
     if (op_ret.data.size()) {
       for (auto it : op_ret.data) {
         cls_rgw_gc_obj_info info;
-        decode(info, it);
+        try {
+          decode(info, it);
+        } catch (buffer::error& err) {
+          CLS_LOG(1, "ERROR: cls_gc_queue_list(): failed to decode gc info\n");
+          return -EINVAL;
+        }
         //Check for info tag in urgent data map
         if (urgent_data_map.size() > 0) {
           auto found = urgent_data_map.find(info.tag);
-          if (found != urgent_data_map.end() && found->second > info.time) {
-            CLS_LOG(1, "INFO: cls_gc_queue_list(): tag found in urgent data: %s\n", info.tag.c_str());
-            continue;
+          if (found != urgent_data_map.end()) {
+            if (found->second > info.time) {
+              CLS_LOG(1, "INFO: cls_gc_queue_list(): tag found in urgent data: %s\n", info.tag.c_str());
+              continue;
+            }
+          }
+        } else {
+          //Search in xattrs
+          bufferlist bl_time;
+          int ret = cls_cxx_getxattr(hctx, info.tag.c_str(), &bl_time);
```

Comment:
XXX When I suggested spilling into xattrs, I imagined that we would potentially use a single xattr to store "additional serialized map data" that we'd decode and search.  The rationale is, avoid degrading to a mode with significant omap loading worst case.  If spillover adds only one well-known key worst case, the -omap- loading per se is minimized?

Reply:
Ok, will make the changes - the value of the xattr key will be an unordered map again which will have tag->expiration time, and the cls_gc interface will have the logic to deserialize it and then search for the tag and get its corresponding time. Does that sound correct now?

Reply:
Is there any limit on the size of the value stored as an xattr?

Reply:
1. ok
2. I don't think so--if we hit it, we might need to write out a sequence of pages or something
Matt

Reply:
per http://docs.ceph.com/docs/master/rados/configuration/filestore-config-ref/, even filestore can deal with arbitrary length xattrs automatically, it might get slower if that part is enormous :)

Reply:
Ok, thanks Matt, will make these changes and all the others that we have discussed.

Reply:
@mattbenjamin: When all the entries are deleted from the unordered map that is saved as an xattr, then should the map and xattr be deleted, and populated again when a new entry spills over to the xattr. Or what is the correct place to delete the xattr of the queue object? Or is this never required?